crucible-llvm/src/Lang/Crucible/LLVM/SimpleLoopFixpoint.hs

```diff
@@ -433,207 +679,365 @@ filterSubstitution sym substitution =
+--   C.CFG ext blocks init ret ->
+--   Integer ->
+--   IO (k, [k])
+-- computeLoopBlocks cfg loopNum =
+--   case List.genericDrop loopNum (Map.toList loop_map) of
+--     [] -> fail ("Did not find " ++ show loopNum ++ " loop headers")
+--     (p:_) -> do checkSingleEntry p
+--                 checkSingleBackedge p
+--                 return p
+
+--  where
+--   -- There should be exactly one block which is not part of the loop body that
+--   -- can jump to @hd@.
+--   checkSingleEntry :: (k,[k]) -> IO ()
+--   checkSingleEntry (hd, body) =
+--     case filter (\x -> not (elem x body) && elem hd (C.cfgSuccessors cfg x)) allReachable of
+--       [_] -> return ()
+--       _   -> fail "SimpleLoopInvariant feature requires a single-entry loop!"
+
+--   -- There should be exactly on block in the loop body which can jump to @hd@.
+--   checkSingleBackedge :: (k,[k]) -> IO ()
+--   checkSingleBackedge (hd, body) =
+--     case filter (\x -> elem hd (C.cfgSuccessors cfg x)) body of
+--       [_] -> return ()
+--       _   -> fail "SimpleLoopInvariant feature requires a loop with a single backedge!"
+
+--   flattenWTOComponent = \case
+--     C.SCC C.SCCData{..} ->  wtoHead : concatMap flattenWTOComponent wtoComps
+--     C.Vertex v -> [v]
+
+--   loop_map = Map.fromList $ mapMaybe
+--     (\case
+--       C.SCC C.SCCData{..} -> Just (wtoHead, wtoHead : concatMap flattenWTOComponent wtoComps)
+--       C.Vertex{} -> Nothing)
+--     wto
+
+--   allReachable = concatMap flattenWTOComponent wto
+
+--   wto = C.cfgWeakTopologicalOrdering cfg
 
 
 -- | This execution feature is designed to allow a limited form of
 --   verification for programs with unbounded looping structures.
 --
---   It is currently highly experimental and has many limititations.
---   Most notibly, it only really works properly for functions
---   consiting of a single, non-nested loop with a single exit point.
+--   It is currently highly experimental and has many limitations.
+--   Most notably, it only really works properly for functions
+--   consisting of a single, non-nested loop with a single exit point.
 --   Moreover, the loop must have an indexing variable that counts up
 --   from a starting point by a fixed stride amount.
 --
---   Currently, these assumptions about the loop strucutre are not
+--   Currently, these assumptions about the loop structure are not
 --   checked.
 --
---   The basic use case here is for verifiying functions that loop
+--   The basic use case here is for verifying functions that loop
 --   through an array of data of symbolic length.  This is done by
 --   providing a \""fixpoint function\" which describes how the live
 --   values in the loop at an arbitrary iteration are used to compute
 --   the final values of those variables before execution leaves the
 --   loop. The number and order of these variables depends on
 --   internal details of the representation, so is relatively fragile.
 simpleLoopFixpoint ::
```

Comment:
This function is an extension/enhancement that does not impact other code and is only utilized when specifically utilizing this fixpoint feature, so while this is messy, it's relatively safe to include to prevent significant bitrot.